Variant Calling Pipeline
========================

1. Required software
====================

* fastqc
* bwa
* STAR
* trimmomatic
* samtools
* picard
* GATK
* python
* Rstudio, gplots,ggplot2,reshape,gsalib


2. Steps before running the pipeline:
=====================================

* run fastqc for one fastq file, if the Encoding version is smaller than 1.8 then trimming is necessary, and phred should be set as 64.


should only change the parameters in block define some parameters:
------------------------------------------------------------------
1). change file_path, Trim, bwaDb, ref_fa, samplename, read_group
2). In file Modules/f08_GATK.py: change -L option in RealignerTarget Creater, HaplotypeCaller,BaseRecalibrator.




Run variant calling pipeline:
=============================
In terminal type: python DNA_vari_call.py

The files generated by Vari_call are:
=====================================

Part I:
-------

1. read files:
--------------
	* fastqfiles filename

2. define group:
----------------
	no files

3. align using bwa:
------------------
	* filename.sam

4. sam2bam_sort:
----------------
	* filename.sort.bam

5. Markduplicates:
------------------
	* filename.dedup.bam

6. Create a target list of intervals:
------------------------------------
	* filename.interval.list

7. Realignment of target intervals
----------------------------------
	* filename.realign.bam

Part II:
--------

1. determine parameters for variant call:
-----------------------------------------
	no file

2. Call variant:
----------------
	* filename.raw.g.vcf

3. JointGenotype:
-----------------
	* samplename.raw.g.vcf

4. Variant Hard filter:
-----------------------

### 1) Select Variants:
	* samplename.raw.snp.vcf
	* samplename.raw.snp.vcf.idx
	* samplename.raw.indel.vcf
	* samplename.raw.indel.vcf.idx
### 2) snp hard filter:
	* samplename.gold.snp.vcf
	* samplename.gold.snp.vcf.idx
### 3) indel hard filter:
	* samplename.gold.indel.vcf
	* samplename.gold.indel.vcf.idx

5. Base recalibration:
----------------------
	* filename.recal.table
	* filename.post_recal.table
	* filename.recal_plots.pdf
	* filename.recal.bam

6. Variant Recalibration:
-------------------------
### 1) build recalibration model for SNP
	* samplename.SNP.recal
	* samplename.tranches
	* samplename.SNP_plots.R
### 2) apply model to SNP
	* sample.recalSnp_rawIndel.vcf
### 3) build recalibration model for indel
	* samplename.indel.recal
	* samplename.indel.tranches
	* samplename.indel.plots.R
### 4) apply model to indel
	* samplename.recal_variant.vcf



3. Annotating the vcf files pipeline:
=====================================
Prepare database for snpEff(it doesn't have one for hamster)
* 0) Add hamster201408.genome : Chinese Hamster to the config file.
* 1) create a folder named data in the snpEff folder.
* 2) In the data folder, create a folder named genomes, another folder named same with that in 0) (hamster201408). 
* 3) Put genome fa file named same with the one in 0). (hamster201408.genome)
* 4) Put annotation file in folder hamster201408 and change its name to genes.gff.
* 5) run the command java -jar snpEff.jar build -gff3 -v hamster201408
* 6) Put all vcf and vcf.idx files into one folder
* 7) Define all parameters in 05_snpEff_Provean_Annotation_Parameters.txt
* 8) In terminal run: nohup python /home/shangzhong/Codes/Pipeline/VariantCall/05_snpEff_Provean_Vari_Annotation.py Annotation_Parameters.txt >log.txt &
